x-pack/plugins/alerting/server/rules_client/methods/clear_expired_snoozes.ts

```diff
@@ -31,18 +26,18 @@ export async function clearExpiredSnoozes(
       })
     : [];
 
-  if (snoozeSchedule.length === attributes.snoozeSchedule?.length) return;
+  if (snoozeSchedule.length === rule.snoozeSchedule?.length) return;
 
   const updateAttributes = updateMeta(context, {
     snoozeSchedule,
     updatedBy: await context.getUserName(),
     updatedAt: new Date().toISOString(),
   });
-  const updateOptions = { version };
+  const updateOptions = { refresh: false };
```

Comment:
**nit:** we may want to keep the `version` passed through to avoid a race condition clearing some data. For example, a user snoozing the rule while the rule is clearing an expired snooze. In that case, it could happen that the clear expired snooze removes the user's requested snooze.

Reply:
Is `version` being sent to the task manager? I was looking through the type definitions but couldn't spot it.

Reply:
I tried doing two things on top of your PR, we may have luck with this approach:
- Move the `clearExpiredSnoozes` closer to when the rule is fetched (to reduce the odds of having the rule updated by something else between read / update)
- Pass through the `version` attribute alongside the `rule` object

Git diff:
```
diff --git a/x-pack/plugins/alerting/server/rules_client/methods/clear_expired_snoozes.ts b/x-pack/plugins/alerting/server/rules_client/methods/clear_expired_snoozes.ts
index 6608521be3a..90156246331 100644
--- a/x-pack/plugins/alerting/server/rules_client/methods/clear_expired_snoozes.ts
+++ b/x-pack/plugins/alerting/server/rules_client/methods/clear_expired_snoozes.ts
@@ -13,7 +13,10 @@ import { updateMeta } from '../lib';
 
 export async function clearExpiredSnoozes(
   context: RulesClientContext,
-  { rule }: { rule: Pick<SanitizedRule<RuleTypeParams>, 'id' | 'snoozeSchedule'> }
+  {
+    rule,
+    version,
+  }: { rule: Pick<SanitizedRule<RuleTypeParams>, 'id' | 'snoozeSchedule'>; version?: string }
 ): Promise<void> {
   const snoozeSchedule = rule.snoozeSchedule
     ? rule.snoozeSchedule.filter((s) => {
@@ -33,7 +36,8 @@ export async function clearExpiredSnoozes(
     updatedBy: await context.getUserName(),
     updatedAt: new Date().toISOString(),
   });
-  const updateOptions = { refresh: false };
+
+  const updateOptions = { version, refresh: false };
 
   await partiallyUpdateAlert(
     context.unsecuredSavedObjectsClient,
diff --git a/x-pack/plugins/alerting/server/rules_client/rules_client.ts b/x-pack/plugins/alerting/server/rules_client/rules_client.ts
index 38cc1f053df..7e2ba8ddcd6 100644
--- a/x-pack/plugins/alerting/server/rules_client/rules_client.ts
+++ b/x-pack/plugins/alerting/server/rules_client/rules_client.ts
@@ -149,6 +149,7 @@ export class RulesClient {
 
   public clearExpiredSnoozes = (options: {
     rule: Pick<SanitizedRule<RuleTypeParams>, 'id' | 'snoozeSchedule'>;
+    version?: string;
   }) => clearExpiredSnoozes(this.context, options);
 
   public muteAll = (options: { id: string }) => muteAll(this.context, options);
diff --git a/x-pack/plugins/alerting/server/task_runner/rule_loader.ts b/x-pack/plugins/alerting/server/task_runner/rule_loader.ts
index 932e98cc535..f006d421fd7 100644
--- a/x-pack/plugins/alerting/server/task_runner/rule_loader.ts
+++ b/x-pack/plugins/alerting/server/task_runner/rule_loader.ts
@@ -38,6 +38,7 @@ export async function loadRule<Params extends RuleTypeParams>(params: LoadRulePa
   let rule: SanitizedRule<Params>;
   let fakeRequest: CoreKibanaRequest;
   let rulesClient: RulesClientApi;
+  let version: string | undefined;
 
   try {
     const attributes = await getRuleAttributes<Params>(context, ruleId, spaceId);
@@ -46,6 +47,7 @@ export async function loadRule<Params extends RuleTypeParams>(params: LoadRulePa
     rule = attributes.rule;
     fakeRequest = attributes.fakeRequest;
     rulesClient = attributes.rulesClient;
+    version = attributes.version;
   } catch (err) {
     throw new ErrorWithReason(RuleExecutionStatusErrorReasons.Decrypt, err);
   }
@@ -85,6 +87,7 @@ export async function loadRule<Params extends RuleTypeParams>(params: LoadRulePa
     apiKey,
     rulesClient,
     validatedParams,
+    version,
   };
 }
 
@@ -99,6 +102,7 @@ export async function getRuleAttributes<Params extends RuleTypeParams>(
   rule: SanitizedRule<Params>;
   fakeRequest: CoreKibanaRequest;
   rulesClient: RulesClientApi;
+  version?: string;
 }> {
   const namespace = context.spaceIdToNamespace(spaceId);
 
@@ -121,6 +125,7 @@ export async function getRuleAttributes<Params extends RuleTypeParams>(
 
   return {
     rule,
+    version: rawRule.version,
     apiKey: rawRule.attributes.apiKey,
     enabled: rawRule.attributes.enabled,
     consumer: rawRule.attributes.consumer,
diff --git a/x-pack/plugins/alerting/server/task_runner/task_runner.ts b/x-pack/plugins/alerting/server/task_runner/task_runner.ts
index b325cd93f59..4cce009355b 100644
--- a/x-pack/plugins/alerting/server/task_runner/task_runner.ts
+++ b/x-pack/plugins/alerting/server/task_runner/task_runner.ts
@@ -491,8 +491,6 @@ export class TaskRunner<
     let executionHandlerRunResult: RunResult = { throttledSummaryActions: {} };
 
     await this.timer.runWithTimer(TaskRunnerTimerSpan.TriggerActions, async () => {
-      await rulesClient.clearExpiredSnoozes({ rule });
-
       if (isRuleSnoozed(rule)) {
         this.logger.debug(`no scheduling of actions for rule ${ruleLabel}: rule is snoozed.`);
       } else if (!this.shouldLogAndScheduleActionsForAlerts()) {
@@ -710,6 +708,18 @@ export class TaskRunner<
       );
       this.ruleMonitoring.setMonitoring(preparedResult.rule.monitoring);
 
+      (async () => {
+        try {
+          await preparedResult.rulesClient.clearExpiredSnoozes({
+            rule: preparedResult.rule,
+            version: preparedResult.version,
+          });
+        } catch (e) {
+          // Most likely a 409 conflict error, which is ok, we'll try again at the next rule run
+          this.logger.debug(`Failed to clear expired snoozes: ${e.message}`);
+        }
+      })();
+
       stateWithMetrics = asOk(await this.runRule(preparedResult));
 
       // fetch the rule again to ensure we return the correct schedule as it may have
```

Reply:
Here's the patch file, I tested it and should work on your end: [approach.patch](https://github.com/elastic/kibana/files/11544478/approach.patch)

Reply:
Changes are looking good, thanks! Let's make sure to have a functional test to ensure the code within `x-pack/plugins/alerting/server/task_runner/task_runner.ts` works and doesn't break in the future (ex: always encountering 409 errors). Let me know once the PR is ready for review (in case you wanted to add more tests, etc).